source/extensions/filters/http/file_system_buffer/filter.cc

```diff
@@ -7,12 +7,459 @@ namespace FileSystemBuffer {
+                                                           Buffer::Instance& data,
+                                                           bool end_stream) {
+  if (state.config_->behavior().bypass()) {
+    return Http::FilterDataStatus::Continue;
+  }
+  state.memory_used_ += data.length();
+  while (data.length() > state.config_->memoryBufferBytesLimit()) {
+    state.buffer_.push_back(
+        std::make_unique<Fragment>(data, state.config_->memoryBufferBytesLimit()));
+  }
+  state.buffer_.push_back(std::make_unique<Fragment>(data));
+  if (end_stream) {
+    state.seen_end_stream_ = true;
+  }
+  dispatchStateChanged();
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+Http::FilterTrailersStatus FileSystemBufferFilter::receiveTrailers(const BufferBehavior& behavior,
+                                                                   BufferedStreamState& state) {
+  // If seen_end_stream_ was populated via receiveData then we shouldn't even get here since that
+  // means there are no trailers. If it was populated by a previous call to receiveTrailers then
+  // that means we must have called continueDecoding or continueEncoding, or we wouldn't get here
+  // again, and we only do that when the data buffer has been flushed, at which point it's time to
+  // deliver any trailers.
+  if (state.seen_end_stream_ || behavior.bypass()) {
+    return Http::FilterTrailersStatus::Continue;
+  }
+  state.seen_end_stream_ = true;
+  dispatchStateChanged();
+  return Http::FilterTrailersStatus::StopIteration;
+}
+
+Http::FilterTrailersStatus FileSystemBufferFilter::encodeTrailers(Http::ResponseTrailerMap&) {
+  return receiveTrailers(response_state_.config_->behavior(), response_state_);
+}
+
+Http::FilterDataStatus FileSystemBufferFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (response_state_.memory_used_ + response_state_.storage_consumed_ + data.length() >
+      response_state_.config_->memoryBufferBytesLimit() +
+          response_state_.config_->storageBufferBytesLimit()) {
+    if (response_state_.injecting_content_length_header_ ||
+        response_state_.config_->behavior().alwaysFullyBuffer()) {
+      response_callbacks_->sendLocalReply(Http::Code::InternalServerError, "buffer limit exceeded",
+                                          nullptr, absl::nullopt, "");
+      return Http::FilterDataStatus::StopIterationNoBuffer;
+    }
+  }
+  return receiveData(response_state_, data, end_stream);
+}
+
+Http::FilterTrailersStatus FileSystemBufferFilter::decodeTrailers(Http::RequestTrailerMap&) {
+  return receiveTrailers(request_state_.config_->behavior(), request_state_);
+}
+
+Http::FilterHeadersStatus FileSystemBufferFilter::encode1xxHeaders(Http::ResponseHeaderMap&) {
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterMetadataStatus FileSystemBufferFilter::encodeMetadata(Http::MetadataMap&) {
+  return Http::FilterMetadataStatus::Continue;
+}
```

Comment:
Nit, if you inherit from PassThroughStreamFilter (I forget exact name) you can avoid this boilerplate but up to you.

Reply:
For such a hefty complex filter I think the "covered everything" is more valuable than the small boilerplate saving.